### PR TITLE
packages/core: genesis snapshots

### PR DESCRIPTION
**Pull request by raykyri**
This is a backwards-compatible release that adds a `Snapshot` message type which can be used for log compaction, app upgrades, and (in the future) putting application code onto the log.

* Adds a `Snapshot` message type, that contains `models` and `effects`. Snapshots may only be applied as the first message in a log.
* Adds an `snapshot` field to `Canvas.initialize({ ... })` which must be provided along with `reset: true`. If provided, it clears the local ModelDB upon initialization and restores the ModelDB models and `$effects` contained in the snapshot, by applying the snapshot as the first message in the log.
* Adds a `createSnapshot` method to packages/core that flattens and packages the current model database as `{ effects: SnapshotEffect[], models: Record<string, SnapshotModelValue> }`.
* Allows gossiplogs to be initialized with a topic of either `xyz.topic` or `xyz.topic#hash` where the hash is hex. Cores should still be initialized with the bare topic `[A-Za-z0-9-.]+` and if a snapshot is provided, a 16-character hash will be appended to the gossiplog topic, which is the hash of the Snapshot.
* Updates the useCanvas hook to detect when a contract has changed, and if so, snapshot the current database and reset it. This will take the current log offline, and will require manually reconnecting.

In the future, we could also snapshot the model schema and action schema (using Function.toString() or posting the contract file). These are planned as optional fields on the `Snapshot` model right now.

A `Commit` message type could be added to the "end" of message logs to signal that canonical execution has moved elsewhere. However, signaling off-the-log may be enough to achieve that.

https://www.notion.so/canvasxyz/DRX-based-on-Snapshots-a4c65ad38ca5490d8ff9cf03e846271d

## Other Questions

- [x] Check that snapshots work with relations
- [x] Check that snapshots work with merge functions
- [x] Tests
- [x] branch and clock are flattened before adding to the snapshot, remove them?
- [x] Snapshot is also a type in gossiplog/src/interface.ts for sync snapshots, rename?
- [x] ~~Should gossiplog topics be `topic/hash` or `topic:hash` instead of `topic#hash`?~~
- [x] verify: are we hashing the snapshot message the same way that gossiplog does hashing?

## How has this been tested?

- [x] CI tests pass
- [x] Tested with snake

## Does this contain any breaking changes to external interfaces?

- [ ] Contract interfaces
- [ ] Core interface
- [ ] CLI
- [ ] Data storage formats, including IndexedDB, SQLite, or filesystem storage (will this break existing apps?)

### CODE REVIEW

File: packages/core/src/runtime/AbstractRuntime.ts [hunk 1]
```diff
@@ -93,30 +86,57 @@ export abstract class AbstractRuntime {
...
+	private static isSnapshot = (message: Message<Action | Session | Snapshot>): message is Message<Snapshot> =>
+		message.payload.type === "snapshot"
+
+	public getConsumer(): GossipLogConsumer<Action | Session | Snapshot> {
 		const handleSession = this.handleSession.bind(this)
 		const handleAction = this.handleAction.bind(this)
+		const handleSnapshot = this.handleSnapshot.bind(this)
 
-		return async function (this: AbstractGossipLog<Action | Session>, signedMessage) {
+		return async function (this: AbstractGossipLog<Action | Session | Snapshot>, signedMessage) {
```
Comment (rjwebb) [inline packages/core/src/runtime/AbstractRuntime.ts:103]: I wonder if we should have a type for `Action | Session | Snapshot`

File: packages/core/src/snapshot.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		// use latest quickjs supported version (es2023)
+		const ast = espree.parse(code, { range: true, ecmaVersion: 2023 })
+		const scopeManager = eslintScope.analyze(ast, { ecmaVersion: 2023 })
+		scopeManager.globalScope.through.forEach((ref) => {
+			if (ref.identifier.name in globals.es2023 || ref.identifier.name in globals.browser) return
+			console.log("unexpected action variable access", ref.identifier.name, "at char", ref.identifier.range?.[0])
+			return true
+		})
+		return false
+	}
```
Comment (rjwebb) [inline packages/core/src/snapshot.ts:77]: What happens if the contract has dangling references?

Comment (raykyri) [reply packages/core/src/snapshot.ts:78]: Nothing at this point; checkContract returns null. But if we start serializing functions canonically inside the Snapshot, then we should probably emit a warning.

**Comment by raykyri** [comment]
There are still a couple of open areas I'm looking at, specifically snapshots with relations, and snapshots with merge functions, but this is ready for review.

**Pull request merged**